File: tests/test_buffer2d.py
import unittest

import aiedit
import numpy as np
import torch


class TestBuffer2D(unittest.TestCase):

    def setUp(self):
        self.raw_data = [[0.4, 0.2], [0.1, 0.7], [0.4, 0.1], [0.3, 0.8], [0.9, 0.2]]
        self.buffer = aiedit.Buffer2D(len(self.raw_data), len(self.raw_data[0]))
        for i in range(self.buffer.num_rows):
            for j in range(self.buffer.num_cols):
                self.buffer[i, j] = self.raw_data[i][j]

    def test_getter(self):
        for i in range(self.buffer.num_rows):
            for j in range(self.buffer.num_cols):
                self.assertAlmostEqual(self.buffer[i, j], self.raw_data[i][j])

    def test_as_numpy_array(self):
        buffer_arr = np.array(self.buffer, copy=False)
        data_arr = np.array(self.raw_data)
        self.assertEqual(buffer_arr.shape, data_arr.shape)
        self.assertTrue(np.allclose(buffer_arr, data_arr))

    def test_as_torch_tensor(self):
        buffer_tensor = torch.from_numpy(np.array(self.buffer, copy=False))
        data_tensor = torch.tensor(self.raw_data)
        self.assertEqual(buffer_tensor.size(), data_tensor.size())
        self.assertTrue(torch.allclose(buffer_tensor, data_tensor))

File: tests/test_buffer2d.cpp
#include <ATen/ops/from_blob.h>
#include <catch2/catch_test_macros.hpp>
#include <vector>

#include "core/buffer2d.hpp"

TEST_CASE("Test setter and getter", "[buffer2d]")
{
    aiedit::Buffer2D buffer(5, 3);
    REQUIRE(buffer.get_num_rows() == 5);
    REQUIRE(buffer.get_num_cols() == 3);
    buffer.set(0, 2, 1.0);
    REQUIRE(buffer.get(0, 2) == 1.0);
    buffer.set(1, 1, 0.5);
    REQUIRE(buffer.get(1, 1) == 0.5);
    REQUIRE(buffer.get(2, 0) == 0.0);
}

TEST_CASE("Test tensor conversion", "[buffer2d]")
{
    std::vector<std::vector<float>> data = {
      {0.2, 0.1, 0.3},
      {0.5, 0.2, 0.8}
    };
    const int num_rows = data.size(), num_cols = data[0].size();
    aiedit::Buffer2D buffer(num_rows, num_cols);
    for (size_t i = 0; i < num_rows; i++) {
        for (size_t j = 0; j < num_cols; j++) {
            buffer.set(i, j, data[i][j]);
        }
    }
    const auto tensor = at::from_blob(buffer.data(), {num_rows, num_cols});
    for (size_t i = 0; i < num_rows; i++) {
        for (size_t j = 0; j < num_cols; j++) {
            REQUIRE(tensor[i][j].item<float>() == data[i][j]);
        }
    }
}